File: 2020/24.py
import time
from utils import readInput
start_time = time.time()

def loadInput():
    lines = readInput("input_24.txt")

    return lines

dirs = {"ne":complex(1, -1), "nw":complex(-1, -1),
        "se":complex(1, 1), "sw":complex(-1, 1),
        "e":complex(2, 0), "w":complex(-2, 0)}
grid = {} 

def part1(lines):
    tiles = 0
    for l in lines:
        i = 0
        pos = complex(0, 0)
        while i < len(l):
            if l[i] == 'e' or l[i] == 'w':
                pos += dirs[l[i]]
                i += 1
            else:
                pos += dirs[l[i:i+2]]
                i += 2

        if pos in grid:
            grid[pos] = 1 - grid[pos]
        else:
            grid[pos] = 1

    print ("🎄 Part 1: {}".format(sum(grid.values())))

def addGrid(grid):
    to_add = {}
    for k in grid.keys():
        neighs = [k + d for d in dirs.values()]
        for n in neighs:
            if n not in grid:
                to_add[n] = 0
    grid.update(to_add)

def part2():
    addGrid(grid)
    for day in range(100):
        to_flip = {}
        for k, v in grid.items():
            val = 0
            for d in dirs.values():
                if k + d in grid:
                    val += grid[k+d]
            if (v == 1 and (val == 0 or val > 2)) or (v == 0 and val == 2):
                to_flip[k] = 1-grid[k]
        grid.update(to_flip)
        addGrid(grid)
    
    print ("🎄🎅 Part 2: {}".format(sum(grid.values())))

print("⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄")
print("⛄        Day 24        ⛄")
print("⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄")

inputs = loadInput()

t0 = time.time()
part1(inputs)
print ("Time: {:.5f}".format(time.time()-t0))

t0 = time.time()
part2()
print ("Time: {:.5f}".format(time.time()-t0))

File: 2020/24.jl
include("utils.jl")
using .AoCUtils: readInput

function loadInput()
    filename = "input_24.txt"
    lines::Vector{String} = readInput(filename) 
    directions_regex = r"e|se|sw|w|nw|ne"
    tiles = Dict{Complex{Int}, Int}() 
    for l in lines
        matches = (m.match for m in eachmatch(directions_regex, l))
        directions = map(x -> dirs[x], matches)
        tiles[sum(directions)] = 1
    end
    return lines
end

function part1(instructions::Vector{String}, dirs::Dict{String, Complex{Int}})
    grid = Dict{Complex{Int}, Int}()

    for ins in instructions
        pos = Complex{Int}(0, 0)
        i = 1
        while i <= length(ins)
            if ins[i] == 'n' || ins[i] == 's'
                d = ins[i:i+1]
                i += 2
            else
                d = string(ins[i])
                i += 1
            end
            pos += dirs[d]
        end

        if haskey(grid, pos)
            grid[pos] = 1 - grid[pos]
        else
            grid[pos] = 1
        end
    end
    
    println("🎄 Part 1: $(sum(values(grid)))")
    return grid
end

function addGrid(grid::Dict{Complex{Int}, Int}, new_grid::Dict{Complex{Int}, Int}, dirs::Dict{String, Complex{Int}})
    for (k, v) in grid
        neighs = [k+d for d in values(dirs)]
        for n in neighs
            if !haskey(grid, n)
                new_grid[n] = 0
            end
        end
    end
    merge!(grid, new_grid)
    empty!(new_grid)
end

function part2(grid::Dict{Complex{Int}, Int}, dirs::Dict{String, Complex{Int}})
    new_grid = Dict{Complex{Int}, Int}()
    addGrid(grid, new_grid, dirs)
    to_flip = Dict{Complex{Int}, Int}()
    for _ in 1:100
        for k in keys(grid)
            val = 0
            for n in [k+d for d in values(dirs)]
                # val += get!(grid, n, 0)
                if haskey(grid, n)
                    val += grid[n]
                else
                    new_grid[n] = 0
                end
            end
            if (grid[k] == 1 && (val == 0 || val > 2)) || (grid[k] == 0 && val == 2)
                to_flip[k] = 1 - grid[k]
            end
        end
        merge!(grid, to_flip)
        merge!(grid, new_grid)
        empty!(to_flip)
        empty!(new_grid)
    end
    println("🎄🎅 Part 2: $(sum(values(grid)))")
end

println("⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄")
println("⛄        Day 24        ⛄")
println("⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄")

dirs = Dict{String, Complex{Int}}("ne"=>1-1im, "nw"=>-1-1im, "se"=>1+1im, 
                                  "sw"=>-1+1im, "e"=>2+0im, "w"=>-2+0im)

inputs = loadInput()
@time grid = part1(inputs, dirs)
@time part2(grid, dirs)
